src/compiler/types.ts

```diff
@@ -603,15 +603,75 @@ namespace ts {
 
     export type DeclarationName = Identifier | StringLiteral | NumericLiteral | ComputedPropertyName | BindingPattern;
 
-    export interface RealDeclaration extends Node {
+    export interface DeclarationBase extends Node {
         _declarationBrand: any;
         name?: DeclarationName;
     }
 
-    // Binary expressions can be declarations if they are 'exports.foo = bar' expressions in JS files
-    export type Declaration = RealDeclaration | BinaryExpression;
-
-    export interface DeclarationStatement extends RealDeclaration, Statement {
+    export type Declaration =
+        | ArrowFunction
+        // Binary expressions can be declarations if they are 'exports.foo = bar' expressions in JS files
+        | BinaryExpression

```

Comment:
So this is for more than just `'exports.foo = bar'` as per the comment? If that's the case, then disregard. I generally haven't been a fan of us tagging `BinaryExpression`, `PropertyAccessExpression`, and `CallExpression` and `NewExpression` as `Declaration`, but that ship has sailed.